ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -214,6 +214,9 @@ func (r *RayClusterReconciler) rayClusterReconcile(request ctrl.Request, instanc
 		if updateErr := r.updateClusterState(instance, rayiov1alpha1.Failed); updateErr != nil {
 			r.Log.Error(updateErr, "RayCluster update state error", "cluster name", request.Name)
 		}
+		if updateErr := r.updateClusterReason(instance, err.Error()); updateErr != nil {
```

Comment:
@DmitriGekhtman The reconcile loop doesn't cease because the `err.Error()` string here is always different. It begins with `pods "POD_NAME" is forbidden: exceeded quota: quota...` and `POD_NAME` keeps changing. So updating the `.status.reason` queues the RayCluster for another reconciliation which updates `.status.reason` which queues it again, on and on.

I'll make the string here stable which should fix.

Reply:
@DmitriGekhtman [this is ugly and brittle](https://github.com/ray-project/kuberay/pull/639/files#diff-cc9abb27aaceca3f10193e2ab35fb00dca44b8858709c5c0f4df751c1387291aR365-R372). Is there a way to have the reconcile loop ignore changes to `.status.reason` or perhaps `.status` altogether? I want to do [something like this](https://github.com/kubernetes-sigs/kubebuilder/issues/618#issuecomment-499411416). I think relevant code is [here](https://github.com/ray-project/kuberay/blob/0579c90d8e807cfa866b175e27beb9afc5526f89/ray-operator/pkg/client/informers/externalversions/ray/v1alpha1/raycluster.go#L51-L55) and/or [here](https://github.com/ray-project/kuberay/blob/0579c90d8e807cfa866b175e27beb9afc5526f89/ray-operator/pkg/client/informers/externalversions/factory.go#L76-L92)? But not sure and need help.

Reply:
Hm, maybe we could just set the reason to something like "Pod reconcile failed" and then display the full error by emitting an event, which would be easily accessible via kubectl describe.

Reply:
Sgtm, I’ll take a look soon

Reply:
Come to think of it, we want to double check that emitting these events doesn't also trigger a tight loop.

Reply:
Aha! [This comment](https://github.com/kubernetes-sigs/kubebuilder/issues/618#issuecomment-595770218) and the one after it on [`GenerationChangedPredicate`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate?utm_source=godoc#GenerationChangedPredicate) are what we need. The latest commit results in correct exponential backoff behavior. [Gist with logs](https://gist.github.com/davidxia/874d87c0ecd5e3d40ce9ac27f474f9fd#file-ray-operator-log) showing gap between timestamp clusters is roughly doubling on each reconcile loop.

[gist showing desired output for `kubectl get` and `describe`](https://gist.github.com/davidxia/874d87c0ecd5e3d40ce9ac27f474f9fd#file-raycluster-output)

Reply:
I'll leave the K8s event work for #670

Reply:
Are there any `.metadata` changes that should trigger reconciliation like label updates?

Reply:
Let's use [Or](https://github.com/kubernetes-sigs/controller-runtime/blob/4dae0f256678f0da2da437a58d013b376acef7bf/pkg/predicate/predicate.go#L291) to also watch for changes to [Labels](https://github.com/kubernetes-sigs/controller-runtime/blob/4dae0f256678f0da2da437a58d013b376acef7bf/pkg/predicate/predicate.go#L218) and [Annotations](https://github.com/kubernetes-sigs/controller-runtime/blob/4dae0f256678f0da2da437a58d013b376acef7bf/pkg/predicate/predicate.go#L187).

CR annotations are currently taken into account by the operator.
Labels are not taken into account for now, but they might be later.

Reply:
The con of ignoring status changes is that extraneous status changes are not quickly fixed. I think that's a minor issue, though.

Reply:
thanks, updated. Also included emitting event since it's small and straightforward. Lmk if there's any good examples of how to test event or label/annotation update-trigger-reconciliation logic.

Reply:
Tested on my cluster. Updating labels or annotations triggers reconciliation as expected. Events also show up as expected. See [gist](https://gist.github.com/davidxia/874d87c0ecd5e3d40ce9ac27f474f9fd#file-events)

Reply:
Nice, thanks!